### PR TITLE
Feature: Shadow DOM for custom elements

### LINKED ISSUE #211: CHORE: include Fiddle build in CI
Our fiddle package is not being built in CI, which allows breaking code to get merged into `main`. We should include it as a build step in the `pr_runner` git CI action

### LINKED ISSUE #233: Custom Elements: support attaching to shadow DOM
Add a feature that can be toggled (e.g. hidden behind a metadata attribute) that will attach a custom element's content to a shadow DOM instead of its `innerHtml`.

### PR DESCRIPTION
**Pull request by samijaber**
## Description

- closes #233 
- closes #211
- adds missing `index` references in loops

### CODE REVIEW

File: packages/core/src/types/mitosis-component.ts [hunk 1]
```diff
@@ -38,6 +38,8 @@ export type MitosisComponent = {
   name: string;
   imports: MitosisImport[];
   meta: JSONObject & {
+    useMetadata?: JSONObject;
+    // TO-DO: remove this? seems to be legacy/outdated, and superseeded by `useMetadata`
     metadataHook?: JSONObject;
```
[the remark this answers is not in the document]
Comment (steve8708) [reply packages/core/src/types/mitosis-component.ts:43]: ah! correct, you can delete the old one. instead of making this a one-off pattern, I instead wanted to move toward an ultimately generalizable pattern where people can make their own metadata hooks and it'll automatically go to the meta object with the name of the hook. 

tl;dr `metadataHook` is not `useMetadata` so you can delete the old name and comment

[the remark this answers is not in the document]
Comment (samijaber) [reply packages/core/src/types/mitosis-component.ts:43]: Sounds good! A small followup:

- it is used in the mitosis generator, and I'm confused by the intent here. Do you have any insight? Can this code be removed as well? 

https://github.com/BuilderIO/mitosis/blob/812b9e042884beb7966e96577f3595df81724de3/packages/core/src/generators/mitosis.ts#L187-L194

- I assume we should correct this in the Stencil generator to use `useMetadata`
 https://github.com/BuilderIO/mitosis/blob/812b9e042884beb7966e96577f3595df81724de3/packages/core/src/generators/stencil/generate.ts#L187

File: packages/core/src/generators/html.ts [hunk 2]
```diff
@@ -679,8 +680,13 @@ export const componentToCustomElement =
           }
 
           ${useOptions.js}
+
+          if (${json.meta.useMetadata?.isAttachedToShadowDom}) {
+            this.attachShadow({ mode: 'open' })
```
Comment (PatrickJS) [inline packages/core/src/generators/html.ts:685]: there is no option for `mode: "close"`
